### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by KhushalPShah**
Contains the base structure of the modules to be used in the library.
Have partially incorporated the changes from the TS library.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+	// create a new internal go-chi router
+	chi_router := chi.NewRouter()
+
+	// read all the routes from the router
+	routes := server.router.routes
+	for _, route := range routes {
+		newFunc := func(w http.ResponseWriter, r *http.Request) {
+			// Decode the JSON request body into the struct instance
+			structPtr := reflect.New(route.handler.inpType).Interface()
```
Comment (YourTechBud) [inline server.go:64]: Move this outside the function

File: router.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+		route = route.fn(function)
+
+	} else if len(items) < 1 || len(items) > 2 {
+		panic("Illegal number of arguments provided to Query")
+	}
+	r.routes = append(r.routes, route)
+	return route
+}
+
+func (r *Router) Mutation(items ...interface{}) *Route {
```
Comment (YourTechBud) [inline router.go:66]: No need to spread parameter for `Mutation` and `Query`. Just accept 1 parameter: `name`.

File: router.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	if len(items) >= 1 {
+		operationId = items[0].(string)
+		route = &Route{
+			handler: &Handler{},
+			routeConfig: &RouteConfig{
+				opId:       operationId,
+				opType:     "mutation",
+				method:     http.MethodPost,
+				url:        "/v1/" + operationId,
+				jsonSchema: &JSONSchemaForInputAndOutput{},
```
Comment (YourTechBud) [inline router.go:80]: I suggest you calculare the jsonSchema here itself

File: router.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func (r *Router) Mutation(items ...interface{}) *Route {
+	var operationId string
+	var function interface{}
+	var route *Route
+	// if the Query function is called with operationID
+	if len(items) >= 1 {
+		operationId = items[0].(string)
+		route = &Route{
```
Comment (YourTechBud) [inline router.go:73]: Use a `NewRoute` func to return a `*Route` object.

File: route.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	output *jsonschema.Schema
+}
+
+type Handler struct {
+	inpType  reflect.Type
+	outType  reflect.Type
+	function interface{}
+}
+
+func (r *Route) method(method string) *Route {
```
Comment (YourTechBud) [inline route.go:36]: Export this method i.e. rename the method to `Method`

File: route.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	// todo - check if the method is valid string
+	r.routeConfig.method = method
+	return r
+}
+func (r *Route) url(url string) *Route {
```
Comment (YourTechBud) [inline route.go:41]: Rename to `URL`

File: route.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	// todo - check for any invalid URL
+	r.routeConfig.url = url
+	return r
+}
+
+func (r *Route) fn(function interface{}) *Route {
```
Comment (YourTechBud) [inline route.go:47]: Rename to `Fn`

File: route.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	inpType, outType, err := validateAndRetrieveHandlerParamType(function)
+	if err != nil {
+		panic(err)
+	}
+	handler := &Handler{inpType, outType, function}
```
Comment (YourTechBud) [inline route.go:52]: Calculate the JSON schema from the `in` and `out` types here itself

File: route.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	noOfInputParam := fnType.NumIn()
+
+	if noOfInputParam == 2 {
+		if reflect.TypeOf((*context.Context)(nil)).Elem() == fnType.In(0) {
+			return fnType.In(1), nil
+		} else {
+			return nil, errors.New("Second input parameter of handler function should be of type context")
+		}
+	} else {
+		return nil, errors.New(fmt.Sprintf("Invalid number of input arguments - Expected : 2, Actual : %v", noOfInputParam))
```
Comment (YourTechBud) [inline route.go:92]: Print a message which suggests the user to pass the `context.Context` object as the first parameter

File: route.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+func validateAndRetrieveOutputParamType(fnType reflect.Type) (reflect.Type, error) {
+
+	noOfOutputParam := fnType.NumOut()
+
+	if noOfOutputParam == 1 {
+		if reflect.TypeOf((*error)(nil)).Elem() == fnType.Out(0) {
+			return fnType.Out(0), nil
+		}
+		return nil, errors.New("If the output of the function has 1 paramteres, it should of type error")
```
Comment (YourTechBud) [inline route.go:105]: Remove this. Let's allow the user to not return any error at all. So the return type may have `0` paramters

File: server.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+		OpenAPI: "3.0.0.",
+		Info: &openapi3.Info{
+			Title:       "Go Server",
+			Description: "List of all the registered endpoints",
+		},
+	}
+
+	for _, route := range routes {
+
+		jsonInpSchema := route.routeConfig.jsonSchema.input
```
Comment (YourTechBud) [inline server.go:184]: Where is this getting calculated??

File: server.go [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+	chi_router := chi.NewRouter()
+
+	// read all the routes from the router
+	routes := server.router.routes
+	for _, route := range routes {
+		newFunc := func(w http.ResponseWriter, r *http.Request) {
+			// Decode the JSON request body into the struct instance
+			structPtr := reflect.New(route.handler.inpType).Interface()
+			decoder := json.NewDecoder(r.Body)
+			decoder.DisallowUnknownFields()
```
Comment (YourTechBud) [inline server.go:66]: This isn't helpful. Use this library instead: https://github.com/xeipuuv/gojsonschema. Note: JSON schema should be pre-calculated in order to use this library

File: server.go [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+	// create a new internal go-chi router
+	chi_router := chi.NewRouter()
+
+	// read all the routes from the router
+	routes := server.router.routes
+	for _, route := range routes {
+		newFunc := func(w http.ResponseWriter, r *http.Request) {
+			// Decode the JSON request body into the struct instance
+			structPtr := reflect.New(route.handler.inpType).Interface()
+			decoder := json.NewDecoder(r.Body)
```
Comment (YourTechBud) [inline server.go:65]: What if it was a `GET` or `DELETE` method?

File: server.go [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+			err := decoder.Decode(structPtr)
+			if err != nil {
+				http.Error(w, "Invalid JSON request body", http.StatusBadRequest)
+				return
+			}
+
+			decodedValue := reflect.ValueOf(structPtr).Elem()
+			decodedType := decodedValue.Type()
+
+			if decodedType != route.handler.inpType {
```
Comment (YourTechBud) [inline server.go:76]: Why is this required?

File: server.go [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+					return
+				}
+			}
+
+			// Fetching the function from the saved Handler
+			fn := route.handler.function
+
+			function := reflect.ValueOf(fn)
+			args := []reflect.Value{
+				reflect.ValueOf(context.TODO()),
```
Comment (YourTechBud) [inline server.go:89]:
```suggestion
				reflect.ValueOf(r.Context()),
```

File: server.go [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+				reflect.ValueOf(structPtr).Elem(),
+			}
+			result := function.Call(args)
+			if len(result) == 1 {
```
Comment (YourTechBud) [inline server.go:93]: lenght can be zero

File: server.go [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+		// Write the JSON file as the response body
+		_, err = w.Write(jsonOpenAPI)
+		if err != nil {
+			// Handle the error, e.g., log it or return an error response
+			http.Error(w, "Failed to send OpenAPI JSON response", http.StatusInternalServerError)
+			return
+		}
+	}
+	chi_router.Get("/openapi.json", endPointForJSONSchemaFunction)
```
Comment (YourTechBud) [inline server.go:166]: Not the same endpoint as me

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: router.go (hunk 2 above)
Comment (KhushalPShah) [reply router.go:66]: In the reference that you had shared with me, you mentioned one use case as- 
router.Mutation("myQuery2", addTodos) 

Hence, the spread was added.

File: router.go (hunk 3 above)
Comment (KhushalPShah) [reply router.go:80]: If the Mutation() or Query() only accepts 1 parameter, i.e. 'name', then I will not get a handler function.
In such case, my routeConfig would have an empty object for jsonSchema.

File: route.go (hunk 10 above)
Comment (KhushalPShah) [reply route.go:105]: So, do you mean that the output can have 0-2 parameters (both inclusive), but nothing beyond these limits?

File: server.go (hunk 11 above)
Comment (KhushalPShah) [reply server.go:184]: I had missed this part.
Will fix.

File: server.go (hunk 12 above)
Comment (KhushalPShah) [reply server.go:66]: Can you please explain why is this not helpful ?

File: server.go (hunk 14 above)
Comment (KhushalPShah) [reply server.go:76]: This is to check if the parsed body in the request matches with the expected Input type for the handler.
But I feel the decoder.Decode() logic above would do that, right?
Hence, this is not needed ?

File: server.go (hunk 15 above)
Comment (KhushalPShah) [reply server.go:89]: Will fix this

File: server.go (hunk 17 above)
Comment (KhushalPShah) [reply server.go:166]: Will fix

File: router.go (hunk 3 above)
Comment (YourTechBud) [reply router.go:80]: Yes. You are correct. Lets calculate the jsonschema in the Fn() method then which will accept the user's function

Comment (KhushalPShah) [reply router.go:80]: Got it. Will do that.

File: route.go (hunk 10 above)
Comment (YourTechBud) [reply route.go:105]: Yes. Possible scenarios:
1. No return type (always send 204 status code in this case
2. Only error: output will fluctuate between 204 and 500
3. Only struct: output will be 200
4. Struct and error: output will be 200 or 500

File: server.go (hunk 12 above)
Comment (YourTechBud) [reply server.go:66]: Json schema based validation includes additional things like:
1. Min and max length of string
2. Enumerations

This type of validation won't be captured with the code currently written.

File: server.go (hunk 14 above)
Comment (YourTechBud) [reply server.go:76]: Yes. It isn't needed
